From: Debian Rust Maintainers <pkg-rust-maintainers@alioth-lists.debian.net>
Date: Sat, 2 Oct 2021 01:08:00 +0100
Subject: d-0001-pkg-config-no-special-snowflake

Description: always enable cross compilation via pkgconf, and set the right binary name.

Forwarded: not-needed
---
 vendor/pkg-config-0.3.30/src/lib.rs    | 26 +++++++++++---------------
 vendor/pkg-config-0.3.30/tests/test.rs |  2 --
 2 files changed, 11 insertions(+), 17 deletions(-)

Index: rustc/vendor/pkg-config-0.3.32/src/lib.rs
===================================================================
--- rustc.orig/vendor/pkg-config-0.3.32/src/lib.rs
+++ rustc/vendor/pkg-config-0.3.32/src/lib.rs
@@ -150,11 +150,8 @@ pub enum Error {
     /// Contains the name of the responsible environment variable.
     EnvNoPkgConfig(String),
 
-    /// Detected cross compilation without a custom sysroot.
-    ///
-    /// Ignore the error with `PKG_CONFIG_ALLOW_CROSS=1`,
-    /// which may let `pkg-config` select libraries
-    /// for the host's architecture instead of the target's.
+    /// Cross compilation detected. Kept for compatibility;
+    /// the Debian package never emits this.
     CrossCompilation,
 
     /// Failed to run `pkg-config`.
@@ -269,14 +266,6 @@ impl fmt::Display for Error {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
         match *self {
             Error::EnvNoPkgConfig(ref name) => write!(f, "Aborted because {} is set", name),
-            Error::CrossCompilation => f.write_str(
-                "pkg-config has not been configured to support cross-compilation.\n\
-                \n\
-                Install a sysroot for the target platform and configure it via\n\
-                PKG_CONFIG_SYSROOT_DIR and PKG_CONFIG_PATH, or install a\n\
-                cross-compiling wrapper for pkg-config and set it via\n\
-                PKG_CONFIG environment variable.",
-            ),
             Error::Command {
                 ref command,
                 ref cause,
@@ -402,7 +391,7 @@ impl fmt::Display for Error {
                 )?;
                 format_output(output, f)
             }
-            Error::__Nonexhaustive => panic!(),
+            Error::CrossCompilation | Error::__Nonexhaustive => panic!(),
         }
     }
 }
@@ -596,6 +585,8 @@ impl Config {
         if host == target {
             return true;
         }
+        // always enable PKG_CONFIG_ALLOW_CROSS override in Debian
+        return true;
 
         // pkg-config may not be aware of cross-compilation, and require
         // a wrapper script that sets up platform-specific prefixes.
@@ -653,7 +644,12 @@ impl Config {
     }
 
     fn run(&self, name: &str, args: &[&str]) -> Result<Vec<u8>, Error> {
-        let pkg_config_exe = self.targeted_env_var("PKG_CONFIG");
+        let pkg_config_exe = self.targeted_env_var("PKG_CONFIG").or_else(|| {
+            self.env_var_os("DEB_HOST_GNU_TYPE").map(|mut t| {
+                t.push(OsString::from("-pkgconf"));
+                t
+            })
+        });
         let fallback_exe = if pkg_config_exe.is_none() {
             Some(OsString::from("pkgconf"))
         } else {
Index: rustc/vendor/pkg-config-0.3.32/tests/test.rs
===================================================================
--- rustc.orig/vendor/pkg-config-0.3.32/tests/test.rs
+++ rustc/vendor/pkg-config-0.3.32/tests/test.rs
@@ -28,7 +28,6 @@ fn find(name: &str) -> Result<pkg_config
     pkg_config::probe_library(name)
 }
 
-#[test]
 fn cross_disabled() {
     let _g = LOCK.lock();
     reset();
@@ -40,7 +39,6 @@ fn cross_disabled() {
     }
 }
 
-#[test]
 fn cross_enabled() {
     let _g = LOCK.lock();
     reset();
